### ISSUE TITLE
Joint Ligand / Protein splitting

### ISSUE THREAD

**Issue by mmagithub** [IBM/Hestia-GOOD#93]
Hello,

Great repository, and thank you for sharing!

I am working with protein-ligand datasets and often need to perform joint splitting that accounts for both protein and ligand similarities simultaneously. This is essential to minimize data leakage, as the models I am building utilize features from both proteins and ligands. Currently, the lack of this feature makes it challenging to create balanced and representative training and testing sets, potentially impacting the performance and generalizability of my models. This issue is particularly pronounced with large datasets (millions of records of SMILES, sequences, and bioactivity values), where brute-force methods are very time-consuming and resource-intensive.

**Possible Implementation**

Introduce a joint splitting feature that employs algorithms to ensure diversity and balance across both protein and ligand dimensions, such as stratified splitting. Providing options for users to specify the proportion of the split and set constraints to maintain the integrity of both datasets would enhance flexibility and usability. This could involve:

Stratified Splitting: Ensuring that both proteins and ligands are evenly represented in the training and testing sets.
User-Defined Parameters: Allowing users to define the split ratio and any specific constraints related to dataset integrity.
Scalability: Optimizing the feature to handle large datasets efficiently without excessive computational overhead.
Alternatives

As an initial approach, implement cluster-based splitting followed by further fine-tuning to maximize the quality of the split and minimize leakages. Additionally, consider creating a separate subset called "Leak" (or a similar name) that can be excluded from model training and hyperparameter tuning but used for final production models. Other alternatives include:

Hierarchical Splitting: First splitting by proteins and then by ligands within each protein subset to ensure controlled distribution.
Manual Splitting Tools: Providing advanced users with tools to manually adjust splits, though this may be time-consuming and error-prone.

Another approach could be to pre-compute similarity matrices for ligands and protein records, and use these pre-computed matrices to aid the splits.

Additional Context

Joint splitting is crucial for tasks like ligand-based screening and drug discovery, where the interaction between proteins and ligands must be accurately represented in both training and testing sets. Providing this feature would streamline workflows and improve the reliability of predictive models in bioinformatics and cheminformatics applications. Efficient and effective dataset splitting can lead to better model performance, reduced overfitting, and more trustworthy predictions in real-world scenarios.

Thank you for considering this feature request.

Best regards,
Marawan

**Comment by RaulFD-creator** [comment]
Hi @mmagithub, this is a really interesting problem to look into. One of my goals for the repository is that all similarity metrics can work with any partitioning algorithm, so I'm inclined for a solution in the lines of the last option.

## Option 1

1. Compute the similarity between proteins and ligands independently (obtaining two matrices).
2. Aggregate those matrices into a single matrix where each cell corresponds to the similarity between two protein-ligand pairs. The aggregation could have multiple (customisable) functions:
   a. Max (or min) functions: max would be the most strict version and I suspect it wouldn't be too useful if the dataset is not sparse.
   b. Weighted sum with a tunable weight (alpha). If alpha = 0, then only protein similarities are considered; if alpha = 1, then only ligand similarities are considered.
3. From these matrix any partitioning strategy can be used (we are working on including other 'weaker' partitioning methods as well).

I like this idea because it's the most consistent with our philosophy so far.

## Option 2
Hierarchical clustering or stratified clustering should be possible to implement as well, but that would fit more into what an expert user can perform and I don't think we should include it into the repository, it would be better, in my opinion as a code snippet, that I can share in this issue or help you develop. The repository contains functions for performing different types of clustering that could be useful for that.

In any case, let me know which of the two options seems the more useful.

**Comment by RaulFD-creator** [comment]
This is an embodiment for **Option 1**, it is a bit different from the rest of the functions in the package, so it may be a while before I integrate it with the rest of the package, but I think it does the trick. If you'd like to try it, let me know if it addresses the issue. I have tested it for the Davis dataset, which is relatively small, for larger datasets it may require significant RAM memory.

```python
import pandas as pd
from hestia.similarity import *
from hestia import HestiaGenerator
from tqdm import tqdm

def multi_entity_similarity(
    df,
    sim_args_ent1,
    sim_fun_ent1,
    sim_args_ent2,
    sim_fun_ent2,
    scale_1
):
    # Precompute unique values and group indices
    df1_field = sim_args_ent1['field_name']
    df2_field = sim_args_ent2['field_name']

    unique_df1 = df.drop_duplicates(df1_field).reset_index(drop=True)
    unique_df2 = df.drop_duplicates(df2_field).reset_index(drop=True)

    df1_to_df = df.groupby(df1_field).apply(lambda g: g.index.to_numpy())
    df2_to_df = df.groupby(df2_field).apply(lambda g: g.index.to_numpy())

    # Compute similarities
    sim_df_1 = sim_fun_ent1(df_query=unique_df1, **sim_args_ent1)
    sim_df_2 = sim_fun_ent2(df_query=unique_df2, **sim_args_ent2)

    # Vectorized construction of query-target-metric relationships
    queries1, targets1, metrics1 = zip(*[
        (q, t, x[2] * scale_1)
        for x in tqdm(sim_df_1.iter_rows(), total=len(sim_df_1))
        for q in df1_to_df[x[0]]
        for t in df1_to_df[x[1]]
    ])
    new_sim_df1 = pl.DataFrame({
        'query': queries1,
        'target': targets1,
        'metric': metrics1
    }, schema={'query': pl.UInt32, 'target': pl.UInt32, 'metric': pl.Float32})
    del queries1, targets1, metrics1
    queries2, targets2, metrics2 = zip(*[
        (q, t, x[2] * (1 - scale_1))
        for x in tqdm(sim_df_2.iter_rows(), total=len(sim_df_2))
        for q in df2_to_df[x[0]]
        for t in df2_to_df[x[1]]
    ])
    new_sim_df2 = pl.DataFrame({
        'query': queries2,
        'target': targets2,
        'metric': metrics2
    }, schema={'query': pl.UInt32, 'target': pl.UInt32, 'metric': pl.Float32})
    del queries2, targets2, metrics2

    # Combine the two DataFrames
    combined_sim_df = pl.concat([new_sim_df1, new_sim_df2])

    # Custom aggregation using agg_fun
    aggregated_sim_df = (
        combined_sim_df.group_by(['query', 'target'])
            .agg(pl.col('metric').sum())
    )
    return aggregated_sim_df


df = pd.read_csv("data/davis.tab", sep='\t')
sim_df = multi_entity_similarity(
    df=df,
    sim_args_ent1={
        "field_name": "X1",
        "fingerprint": "ecfp",
        "radius": 2,
        "threshold": 0.3
    },
    sim_fun_ent1=molecular_similarity,
    sim_args_ent2={
        "field_name": "X2",
        "threshold": 0.3
    },
    sim_fun_ent2=sequence_similarity_mmseqs,
    scale_1=0.7
)

hg = HestiaGenerator(df)
hg.calculate_partitions(
    sim_df=sim_df, threshold_step=0.1
)
for th, v in hg.get_partitions(filter=0.185):
    print(f"Viable threshold: {th}")
```

**Comment by RaulFD-creator** [comment]
I see. Scaling up to those sizes will definitely be an issue. I will investigate whether I can come up with any solution and update this issue. It may take some time

**Comment by mmagithub** [comment]
Hi Raul. Any updates on this thread ?

**Comment by RaulFD-creator** [comment]
Hi @mmagithub, I'm still running some experiments. Currently, the idea that I have is:

1. Partition train/test based on the proteins (the targets).
2. For the test set, partition it into a loose test (with ligands that may also appear in training) and strict test (with unique ligands).

The main issue I'm having is that computing the fingerprints for all the molecules of the dataset occupies too much memory, so I'm trying to find a way to reduce the memory footprint. I am also investigating using some of the metrics like [iSim](https://github.com/mqcomplab/iSIM) to provide a more efficient computation, though the partitions may be a bit looser. 

Thanks for your patience, this problem is quite interesting. As soon as I have something working I'll get back to you; and if you have any more thoughts/suggestions I'm happy to hear them

**Comment by mmagithub** [comment]
No worries, take your time, thanks Raul.

**Comment by RaulFD-creator** [comment]
Hi @mmagithub, it is a bit of a rough approach, but I have in this [notebook](https://github.com/IBM/Hestia-GOOD/blob/rfd-joint/examples/joint_sim.ipynb), an example of the implementation described above. It doesn't take too long to run (2h approx) and should not require too much RAM.

Main drawback is that if you want to create multiple train/test partitions you need to run it multiple times modifying the thresholds. Another drawback is that the molecule strict test is quite small 5% of the test set (which is 20% of the total data). I'm using a fairly restrictive threshold, 0.3, so with a more lenient one you may get a bit bigger set. Also I'm running the test with a fingerprint with 1,024 bits, just to make the testing easier; but with a larger number of bits you might get a bit better separation.

Feel free to try it yourself, and if you run into any problems, or want clarification let me know. The code is quite rough, I hope you do not mind it too much, I still want to run some more experiments, and after I'll update it. If you have any thoughts on how to improve it, make it clearer, they are more than welcome.

**Comment by mmagithub** [comment]
Thanks Raul. Running now, I will let you know if it works. Thanks

**Comment by mmagithub** [comment]
By the way, for similar problems, people usually tend to be more permissive with ligand similarity thresholds, that protein similarity thresholds, as leakage here, if it happens, is a  joint leakage not necessary attributed to ligand similarities, which is always the case in traditional ligand only data splits

**Comment by mmagithub** [comment]
Hi Raul,

It worked. I just needed to implement an extra safeguard step. Essentially, I took the output from Hestia—which uses MMSeqs2 as a heuristic—and applied a more stringent full alignment method (using Needleman–Wunsch in Biopython) for additional safety. While this extra layer isn’t absolutely necessary, I feel more comfortable applying it to the subsets I'll use for hyperparameter optimization (since I'll eventually use the full dataset).

To illustrate, I generated four histograms showing the distribution of ligand versus sequence similarities/identities, both before and after leak removal. In this context, "leaking records" are defined as those in the training set whose maximum identity to any record in the test or validation sets exceeds 40% for sequence identity or 30% for ligand similarity.

Let me know what you think about these results. Note that this is just a sample—the full leak flagging for the entire dataset is currently running (so there shouldn’t be two distinct distributions). The downside, however, is that the training sets were aggressively reduced—by roughly 10-fold.

Looking forward to your thoughts.

![Image](https://github.com/user-attachments/assets/c65f3515-a180-4eac-bc14-ee28732c6d81)
![Image](https://github.com/user-attachments/assets/db8491b2-512e-46fc-ba68-5e133c28060b)
![Image](https://github.com/user-attachments/assets/285fe71c-13c2-40c2-8ff2-e9c4530ef25a)
![Image](https://github.com/user-attachments/assets/ec2b973c-4275-4342-aee4-c00fdc6dfc84)

**Comment by RaulFD-creator** [comment]
Hi @mmagithub, I think it is a great idea. However, if you don't mind the extra computational burden of Needleman-Wunsch to the whole dataset, you can try to recreate the train/test split step substituting 'mmseqs' in the notebook for 'needle' as the similarity function. That way, the partitioning algorithm will try to find the necessary clusters to create testing, without need to remove any data point. If there are not enough data points, you could always reduce the test size from 0.2 to 0.1. That way you don't lose information. If you have the PDB of the proteins (or their AlphaFold predictions) you could also use Foldseek to compute structural alignments.

All these things are more computationally expensive, your approach on the other hand, if I understand correctly would reduce the number of global alignments necessary.

A final note, the Needleman-Wunsch implementation that the package uses is EMBOSS, I'm not sure whether is the same that Biopython uses; but they should behave similarly.

**Comment by RaulFD-creator** [comment]
Closing as I think the issue has been resolved!
